crates/pathfinder/src/rpc/v02/method/get_state_update.rs

```diff
@@ -0,0 +1,477 @@
+        BlockId::Hash(hash) => hash.into(),
+        BlockId::Number(number) => number.into(),
+    };
+
+    let storage = context.storage.clone();
+    let span = tracing::Span::current();
+
+    let jh = tokio::task::spawn_blocking(move || {
+        let _g = span.enter();
+        let mut db = storage
+            .connection()
+            .context("Opening database connection")?;
+
+        let tx = db.transaction().context("Creating database transaction")?;
+
+        let block_hash = match block_id {
+            StarknetBlocksBlockId::Hash(h) => h,
+            StarknetBlocksBlockId::Number(_) | StarknetBlocksBlockId::Latest => {
+                StarknetBlocksTable::get_hash(
+                    &tx,
+                    block_id.try_into().expect("block_id is not a hash"),
+                )
+                .context("Read block from database")?
+                .ok_or(GetStateUpdateError::BlockNotFound)?
+            }
+        };
+
+        let state_update = StarknetStateUpdatesTable::get(&tx, block_hash)
+            .context("Read state update from database")?
+            .ok_or(GetStateUpdateError::BlockNotFound)?;
+
+        Ok(state_update.into())
+    });
+
+    jh.await.context("Database read panic or shutting down")?
+}
+
+mod types {
+    use serde::Serialize;
+    use serde_with::skip_serializing_none;
+
+    use std::collections::HashMap;
+
+    use crate::core::{
+        ClassHash, ContractAddress, ContractNonce, GlobalRoot, StarknetBlockHash, StorageAddress,
+        StorageValue,
+    };
+
+    #[skip_serializing_none]
+    #[derive(Clone, Debug, Serialize, PartialEq, Eq)]
+    #[serde(deny_unknown_fields)]
+    pub struct StateUpdate {
+        /// None for `pending`
+        #[serde(default)]
+        pub block_hash: Option<StarknetBlockHash>,
+        pub new_root: GlobalRoot,
+        pub old_root: GlobalRoot,
+        pub state_diff: StateDiff,
+    }
+
+    impl From<crate::sequencer::reply::StateUpdate> for StateUpdate {
+        fn from(x: crate::sequencer::reply::StateUpdate) -> Self {
+            Self {
+                block_hash: x.block_hash,
+                new_root: x.new_root,
+                old_root: x.old_root,
+                state_diff: x.state_diff.into(),
+            }
+        }
+    }
+
+    impl From<crate::rpc::v01::types::reply::StateUpdate> for StateUpdate {
+        fn from(x: crate::rpc::v01::types::reply::StateUpdate) -> Self {
+            Self {
+                block_hash: x.block_hash,
+                new_root: x.new_root,
+                old_root: x.old_root,
+                state_diff: x.state_diff.into(),
+            }
+        }
+    }
+
+    /// L2 state diff.
+    #[derive(Clone, Debug, Serialize, PartialEq, Eq)]
+    #[cfg_attr(any(test, feature = "rpc-full-serde"), derive(serde::Deserialize))]
+    #[serde(deny_unknown_fields)]
+    pub struct StateDiff {
+        pub storage_diffs: Vec<StorageDiff>,
+        pub declared_contracts: Vec<DeclaredContract>,
```

Comment:
I think this is incorrect - should be [`declared_contract_hashes`](https://github.com/starkware-libs/starknet-specs/blob/v0.2.0/api/starknet_api_openrpc.json#L869-L875).
```suggestion
pub declared_contract_hashes: Vec<ClassHash>,
```

Reply:
That's a good catch!

I've also added a JSON fixture so that it's easier to see what the input/output of the serialization of the `StateUpdate` structure is (and if it matches the spec).

Reply:
Yeah I had to squint pretty hard :D Really makes me wish we had that flattening tool.